controller_manager/src/controller_manager.cpp

```diff
@@ -65,48 +72,541 @@ ControllerManager::load_controller(
+  }
+
+  if (!robot_hw_->prepareSwitch(switch_start_list_, switch_stop_list_)) {
+    RCLCPP_ERROR(
+      get_logger(),
+      "Could not switch controllers. The hardware interface combination "
+      "for the requested controllers is unfeasible.");
+    stop_request_.clear();
+    start_request_.clear();
+    return controller_interface::return_type::ERROR;
+  }
+#endif
+
+  if (start_request_.empty() && stop_request_.empty()) {
+    RCLCPP_INFO(get_logger(), "Empty start and stop list, not requesting switch");
+    return controller_interface::return_type::SUCCESS;
+  }
+
+  // start the atomic controller switching
+  switch_params_.strictness = strictness;
+  switch_params_.start_asap = start_asap;
+  switch_params_.init_time = rclcpp::Clock().now();
+  switch_params_.timeout = timeout;
+  switch_params_.do_switch = true;
+
+
+  // wait until switch is finished
+  RCLCPP_DEBUG(get_logger(), "Request atomic controller switch from realtime loop");
+  while (rclcpp::ok() && switch_params_.do_switch) {
+    if (!rclcpp::ok()) {
+      return controller_interface::return_type::ERROR;
+    }
+    std::this_thread::sleep_for(std::chrono::microseconds(100));
+  }
+  start_request_.clear();
+  stop_request_.clear();
+
+  RCLCPP_DEBUG(get_logger(), "Successfully switched controllers");
+  return controller_interface::return_type::SUCCESS;
+}
+
 std::shared_ptr<controller_interface::ControllerInterface>
 ControllerManager::add_controller_impl(
   std::shared_ptr<controller_interface::ControllerInterface> controller,
   const std::string & controller_name)
 {
+  // get reference to controller list
+  int free_controllers_list = (current_controllers_list_ + 1) % 2;
+  while (rclcpp::ok() && free_controllers_list == used_by_realtime_) {
+    if (!rclcpp::ok()) {
+      return nullptr;
+    }
+    std::this_thread::sleep_for(std::chrono::microseconds(200));
+  }
+  std::vector<std::shared_ptr<controller_interface::ControllerInterface>>
+  & from = controllers_lists_[current_controllers_list_],
+  & to = controllers_lists_[free_controllers_list];
+  to.clear();
+
+  // Copy all controllers from the 'from' list to the 'to' list
+  for (const auto & controller : from) {
+    to.push_back(controller);
+  }
+
+  // Checks that we're not duplicating controllers
+  for (const auto & controller : to) {
+    if (controller->get_lifecycle_node()->get_name() == controller_name) {
+      to.clear();
+      RCLCPP_ERROR(
+        get_logger(),
+        "A controller named '%s' was already loaded inside the controller manager",
+        controller_name.c_str());
+      return nullptr;
+    }
+  }
+
   controller->init(hw_, controller_name);
+
+  // TODO(v-lopez) this should only be done if controller_manager is configured.
+  // Probably the whole load_controller part should fail if the controller_manager
+  // is not configured, should it implement a LifecycleNodeInterface
+  controller->get_lifecycle_node()->configure();
```

Comment:
@Karsten1987 Could I have your opinion regarding this? I do not fully understand why some classes inherit from `Nodes`, others from `LifeCycleNode` and others from `LifeCycleNodeInterface`, but have a member `LifeCycleNode`.

And to the best of my Google skills there's not much written about when to use what.